### PR TITLE
feat(command): new command: uniq()

### PR DESCRIPTION
**Pull request by joshi-sh**
The [uniq](http://manpages.ubuntu.com/manpages/precise/en/man1/uniq.1.html) utility removes adjacent identical lines from it's input. I have implemented uniq with support for the -i (ignore case) option.

### CODE REVIEW

**Comment by ariporad** [comment]
@nfischer: What are your thoughts on this? I'm not really sure this needs to be a part of shelljs, this could maybe be a plugin, or just a js lib... It's not really shell specific.

**Comment by nfischer** [comment]
@ariporad This is part of GNU core-utils ([here's the source for that](https://www.gnu.org/software/coreutils/manual/coreutils.html#uniq-invocation)), so it seems pretty shell-specific to me. What do you think?

I think this could be a pretty useful utility for some people.

@DaedalusUsedPerl Could you please update this to fix the lint errors?

File: test/uniq.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    fs = require('fs');
+
+shell.config.silent = true;
+
+shell.rm('-rf', 'tmp');
+shell.mkdir('tmp');
+
+var result;
+
+result = shell.uniq();
```
Comment (nfischer) [inline test/uniq.js]: Please mark this section with:

```
//
// Invalids
//
```

See [here](https://github.com/shelljs/shelljs/blob/master/test/cat.js#L11-13) as an example.

File: test/uniq.js [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+assert.ok(shell.error());
+assert.ok(result.code);
+
+assert.equal(fs.existsSync('/asdfasdf'), false); // sanity check
+result = shell.sort('/adsfasdf'); // file does not exist
+assert.ok(shell.error());
+assert.ok(result.code);
+
+//uniq file1
```
Comment (nfischer) [inline test/uniq.js]: Please mark this section with:

```
//
// Valids
//
```

See [here](https://github.com/shelljs/shelljs/blob/master/test/cat.js#L26-28) as an example.

File: test/uniq.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+assert.equal(shell.error(), null);
+assert.equal(result.code, 0);
+assert.equal(shell.cat('resources/uniq/file1u').trimRight(), 
+             shell.cat('resources/uniq/file1t').trimRight());
+             
+//cat file1 |uniq
+result = shell.cat('resources/uniq/file1').uniq();
+assert.equal(shell.error(), null);
+assert.equal(result.code, 0);
+assert.equal(result + '', shell.cat('resources/uniq/file1u').trimRight() + '\n');
```
Comment (nfischer) [inline test/uniq.js]: Why do we need `.trimRight()` here? If this is an issue with ShellStrings being slightly different than regular strings, than you could probably get away with:

`shell.cat('resources/uniq/file1u').toString()`

File: src/uniq.js [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+  uniqed = uniqed.
+             //Do we want only duplicated objects?
+             filter(function(obj){return options.duplicates ? obj.count > 1 : true;}).
+             //Are we tracking the counts of each line?
+             map(function(obj){return (options.count ? (lpad(7,obj.count) + " ") : "") + obj.ln;}).
+             join('\n') + '\n';
+
+  var res = new common.ShellString(uniqed, common.state.error, common.state.errorCode);
+  if(output){
+      res.to(output);
```
Comment (nfischer) [inline src/uniq.js]: Does anything get returned in this branch?

File: src/uniq.js [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    common.error('no input given');
+
+  var lines = (input ? fs.readFileSync(input, 'utf8') : pipe).
+              trimRight().
+              split(/\r*\n/);
+
+  //Perform a run-length encoding of the lines
+  var uniqed = [{count: 1, ln: lines[0]}];
+  lines.slice(1).forEach(function(line){
+      var cmp = options.ignoreCase ?
```
Comment (nfischer) [inline src/uniq.js]: Could you please update this to 2 spaces for indentation? Looks like it's 4 here. Thanks!

File: README.md [hunk 6]
```diff
@@ -405,6 +405,24 @@ Return the contents of the files, sorted line-by-line. Sorting multiple
 files mixes their content, just like unix sort does.
 
 
+### uniq([options,] [input, [output]])
+Available options:
+
++ `-i`: Ignore differences in case when comparing
++ `-c`: Prefix lines by the number of occurrences
```
Comment (nfischer) [inline README.md]: This sounds a bit ambiguous. Could you reword this?

File: README.md [hunk 7]
```diff
@@ -405,6 +405,24 @@ Return the contents of the files, sorted line-by-line. Sorting multiple
 files mixes their content, just like unix sort does.
 
 
+### uniq([options,] [input, [output]])
+Available options:
+
++ `-i`: Ignore differences in case when comparing
```
Comment (nfischer) [inline README.md]: Maybe just, "ignore case while comparing"?

File: README.md [hunk 8]
```diff
@@ -405,6 +405,24 @@ Return the contents of the files, sorted line-by-line. Sorting multiple
...
++ `-c`: Prefix lines by the number of occurrences
++ `-d`: Only print duplicate lines, one for each group
```
Comment (nfischer) [inline README.md]: It's unclear what a "group" is here

File: shell.js [hunk 9]
```diff
@@ -94,6 +94,10 @@ exports.sed = common.wrap('sed', _sed, {idx: 3, canReceivePipe: true}); // don't
 var _sort = require('./src/sort');
 exports.sort = common.wrap('sort', _sort, {idx: 1, canReceivePipe: true});
 
+//@include ./src/uniq
+var _uniq = require('./src/uniq');
+exports.uniq = common.wrap('uniq', _uniq, {idx: 1, canReceivePipe: true});
```
Comment (nfischer) [inline shell.js]: Thanks for the refactor!

File: src/uniq.js [hunk 10]
```diff
@@ -0,0 +1,76 @@
+var common = require('./common');
+var fs = require('fs');
+
+//add c spaces to the left of str
+function lpad(c, str){
+    var res = "" + str;
```
Comment (nfischer) [inline src/uniq.js]: Could you make this an indent of 2 spaces please?

File: src/uniq.js [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+    if(res.length < c){
+        res = Array((c-res.length)+1).join(" ") + res;
```
Comment (nfischer) [inline src/uniq.js]: nit: we prefer single-quoted strings over double-quoted strings when convenient

File: src/uniq.js [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+  var pipe = common.readFromPipe(this);
+
+  if (!input && !pipe)
+    common.error('no input given');
+
+  var lines = (input ? fs.readFileSync(input, 'utf8') : pipe).
+              trimRight().
+              split(/\r*\n/);
+
+  //Perform a run-length encoding of the lines
```
Comment (nfischer) [inline src/uniq.js]: What does this comment mean? It'd be better if it was a bit more descriptive.

File: src/uniq.js [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+  var uniqed = [{count: 1, ln: lines[0]}];
+  lines.slice(1).forEach(function(line){
+    var cmp = options.ignoreCase ? 
+                line.toLocaleLowerCase().localeCompare(uniqed[uniqed.length-1].ln.toLocaleLowerCase()) :
```
Comment (nfischer) [inline src/uniq.js]: Is there a reason this uses `toLocaleLowerCase` vs plain `toLowerCase`? If so, that's fine. I just don't want any unexpected results down the road.

Comment (nfischer) [inline src/uniq.js]: Can we move `uniqed[uniqed.length-1].ln` out of the loop?

File: src/uniq.js [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+                line.localeCompare(uniqed[uniqed.length-1].ln);
+    if(cmp !== 0){
+      uniqed.push({count: 1, ln: line});
+    }else{
+      uniqed[uniqed.length - 1].count++;
+    }
+  });
+  uniqed = uniqed.
+             //Do we want only duplicated objects?
+             filter(function(obj){return options.duplicates ? obj.count > 1 : true;}).
```
Comment (nfischer) [inline src/uniq.js]: Could you write out callbacks in the form:

```javascript
filter(function (obj) {
  // Do we want only duplicated objects?
  return options.duplicates ? obj.count > 1 : true;
}).map( // etc ...
```

**Comment by nfischer** [comment]
@joshi-sh I left some comments. Overall, looks pretty thorough! This is something I'd really like to see in ShellJS. Most of the comments are very nitpicky, so I can fix them if you'd like.

File: src/uniq.js (hunk 13 above)
Comment (joshi-sh) [inline src/uniq.js]: Just to keep it consistent with `sort`, which uses `localeCompare`.

Comment (nfischer) [inline src/uniq.js]: @joshi-sh Yeah, I know it uses `localeCompare`, but why `toLocaleLowerCase`?

**Comment by joshi-sh** [comment]
Your comments are pretty reasonable, actually. I just copied the GNU man pages in lieu of writing docs.

File: src/uniq.js (hunk 13 above)
Comment (joshi-sh) [inline src/uniq.js]: Some locales (like Turkish) don't follow the usual case mappings in Unicode. In most locales, it behaves exactly the same as `toLowerCase`.

**Comment by nfischer** [comment]
Gonna give this an LGTM. We might want to wait to merge until after shelljs v0.8.1, and then this can be in 0.8.2.

**Comment by nfischer** [comment]
LGTM :+1:

**Comment by nfischer** [comment]
I'll handle merging this once we get a patch release out with some of the fixes we've made since v0.7.0.

**Pull request merged**